### PR TITLE
Improve `BIGNUM` operations

### PR DESCRIPTION
**Pull request by Nashtare**
Amortize calls to `modexp` precompile by ~5%.

### CODE REVIEW

File: evm/src/cpu/kernel/asm/bignum/modmul.asm [hunk 1]
```diff
@@ -21,28 +21,34 @@ global modmul_bignum:
...
+    PUSH @SEGMENT_KERNEL_GENERAL
+    GET_CONTEXT
+    %build_address_no_offset
+
     PUSH 0
-    // stack: i=0, len, a_loc, b_loc, m_loc, out_loc, s1, s2, s3, retdest
-modmul_remainder_loop:
-    // stack: i, len, a_loc, b_loc, m_loc, out_loc, s1, s2, s3, retdest
+    // stack: i=0, base_addr, len, a_loc, b_loc, m_loc, out_loc, s1, s2, s3, retdest
+global modmul_remainder_loop:
```
Comment (hratoanina) [inline evm/src/cpu/kernel/asm/bignum/modmul.asm:30]: Is the `global` needed?

File: evm/src/cpu/kernel/asm/bignum/modmul.asm [hunk 2]
```diff
@@ -130,33 +142,38 @@ modmul_return_4:
     // STEP 6:
     // Check that x + k * m = a * b.
 
-    // Walk through scratch_2 and scratch_3, checking that they are equal.
```
Comment (hratoanina) [inline evm/src/cpu/kernel/asm/bignum/modmul.asm:133]: Any reason why the comment is deleted?

File: evm/src/cpu/kernel/asm/bignum/modmul.asm [hunk 3]
```diff
@@ -130,33 +142,38 @@ modmul_return_4:
...
-    SWAP1
-    // stack: mem[i], mem[j], n, i, j, retdest
+    // stack: base_addr, n, i, j, retdest
+    %stack (addr, l, i, j) -> (j, i, addr, addr, l, i, j)
+    // stack: j, i, base_addr, base_addr, n, i, j, retdest
+    DUP3 ADD // addr_j
+    MLOAD_GENERAL
+    // stack: mem[j], i, base_addr, base_addr, n, i, j, retdest
+    SWAP2
+    ADD // addr_j
```
Comment (hratoanina) [inline evm/src/cpu/kernel/asm/bignum/modmul.asm:157]:
```suggestion
    ADD // addr_i
```

File: evm/src/cpu/kernel/asm/bignum/mul.asm [hunk 4]
```diff
@@ -12,49 +12,56 @@ global mul_bignum:
...
-    DUP5
-    // stack: a_start_loc, b[i], mul_addmul_return, n, len, a_start_loc, bi, output_cur, retdest, b
-    DUP8
-    // stack: output_loc, a_start_loc, b[i], mul_addmul_return, n, len, a_start_loc, bi, output_cur, retdest, b
+    // stack: mul_addmul_return, n, base_addr, len, a_start_loc, bi, output_cur, retdest
+    DUP6
+    // stack: bi, mul_addmul_return, n, base_addr, len, a_start_loc, bi, output_cur, retdest
+    DUP4 ADD // bi_addr
+    MLOAD_GENERAL
+    // stack: b[i], mul_addmul_return, n, base_addr, len, a_start_loc, bi, output_cur, retdest, b
```
Comment (hratoanina) [inline evm/src/cpu/kernel/asm/bignum/mul.asm:30]: Unrelated to the PR, but what's `b` at the end of the stack?

**Review by hratoanina — APPROVED** [review]
Looks good, thanks!

File: evm/src/cpu/kernel/asm/bignum/modmul.asm (hunk 1 above)
Comment (Nashtare) [reply evm/src/cpu/kernel/asm/bignum/modmul.asm:30]: for debugging purposes only :upside_down:

File: evm/src/cpu/kernel/asm/bignum/mul.asm (hunk 4 above)
Comment (Nashtare) [reply evm/src/cpu/kernel/asm/bignum/mul.asm:30]: no idea, but it's after retdest, so technically unrelevant to this section (probably a typo anyway). I removed it.

File: evm/src/cpu/kernel/asm/bignum/modmul.asm (hunk 2 above)
Comment (Nashtare) [reply evm/src/cpu/kernel/asm/bignum/modmul.asm:133]: Not sure why it appears deleted, it's still there 🤔

**Review by muursh — APPROVED** [review]
Looks fine to me

**Comment by Nashtare** [comment]
@hratoanina I'll let you judge if you wanna do another pass for the last commit after your review that touches a bit everything due to the new convenience macro, otherwise if it looks good to you I'll merge it

**Pull request merged**